### ISSUE TITLE
NumPy with MKL 2023.1.0 causes RI-K integral polluted for some small molecule with large auxbasis

### ISSUE THREAD

**Issue by ajz34** [pyscf/pyscf#2004]
Hi devs!

## Main conclusion

Probably I found a very tricky problem of MKL, which causes PySCF also break down on this specific task of RI-K integral evaluation.

My analysis show that it's a potential bug of MKL 2023.1.0, but no decisive evidence to support this conjecture.

## Important notes

I think this is not a bug of PySCF, but MKL 2023.1.0 (linked with numpy) instead, after some debugging efforts. But since this problem affected PySCF, so I try to raise issue here.

And since this is not a bug of PySCF, I'm not aware whether this issue should be open or closed. So close this issue if appropriate.

This problem is really tricky. Conditions to reproduce this problem is very restrictive.

## Problem description

This problem was originated when I was tring to evaluate CP-SCF equation of water molecule.

Given zero density matrix, RI-K may not be zero. Code to reproduce:
```python
# problem.py
from pyscf import gto, scf, df, lib
import numpy as np

lib.num_threads(8)

if __name__ == '__main__':
    mol = gto.Mole(atom="""
        O 0.0000 0.0000 0.1173
        H 0.0000 0.7572 -0.4692
        H 0.0000 -0.7572 -0.4692
    """, basis='aug-cc-pVDZ').build()
    mf = scf.RHF(mol).density_fit(auxbasis=df.aug_etb(mol)).run()
    nao = mol.nao
    print(mf.get_k(dm=np.zeros((nao, nao))))
```
All-zero output is expected. However, output matrix may be someting like the following (containing non-zero, random, seems-to-be-uninitialized values).
```
[[ 0.00000000e+000  0.00000000e+000  0.00000000e+000 ...  6.96692393e+199
   2.53064203e-258  2.25563609e-153]
 [ 1.88652445e-317  1.80187717e-318  0.00000000e+000 ...  1.80960100e-094
   1.06112435e-153  8.36665325e+106]
 [ 6.99440989e-283  4.58385649e-278  1.45511200e-296 ...  1.16294398e-012
   2.42766139e-154  8.54523262e+194]
 ...
 [ 9.74455467e-005  0.00000000e+000  6.02518450e-004 ...  3.99190188e+252
   1.87673453e-152  4.23661523e+175]
 [ 2.37259848e-007  3.43258772e-005  4.15705159e-014 ...  5.26750787e+170
   1.51785654e-152  0.00000000e+000]
 [-1.75932942e-004 -2.50561795e-004  0.00000000e+000 ...  0.00000000e+000
   0.00000000e+000  3.79133423e-062]]
```

Notes of reproduction of problem:
- Linux x86-64. Reproduced in more than one servers/clusters. PySCF 2.4.0 or master, locally compiled.
- Number of threads should be explicitly set (either by `lib.num_threads` or `OMP_NUM_THREADS`), and must be **no less than 8 physical cores**. So to reproduce this problem, a CPU with more than 8 physical cores is required. 
- This problem **only occurs on Intel CPU** (tested: E5-2673, E5-2682, Gold 6154, Platinum 9282, ecs.c5.4xlarge of aliyun, ecs.t5-c1m1.4xlarge (Platinum 8163) of aliyun).
- This problem **would not occur on AMD CPU** (tested: Ryzen 7945-HX, EPYC 7642, EPYC 7742).
- This problem is **related to MKL version linked with numpy**. Problem only occurs when mkl=2023.1.0 (`h213fc3f_46344`) linked with numpy=1.26.2 (`py310h5f9d8c6_0`) shipped with conda channel `defaults`. Other versions of mkl (2023.0.0 / 2022.1.0 with `defaults`) or OpenBLAS (shipped by pypi or conda channel `conda-forge`) will not show problem.
- This problem also occurs on **very few cases** with similar molecular size (like SH2, PH2), basis-set (aug-pc-1) and auxiliary-basis-set (`df.aug_etb` with not very large `beta`). Using proper auxiliary-basis-set (aug-cc-pVDZ-jkfit) may not show problem; however, the cause of problem is unrelated to selection of auxiliary-basis-set.
- This problem could be traced to `lib.dot` or `NPdgemm`, but calling `lib.dot` only without running other tasks may not show this problem. In other word, `.run()` in the code above is essential to reproduce this problem.
- This problem is **not related to version of BLAS when compiling PySCF** (pyscf/lib/CMakeLists.txt; libxc and xcfun is not required to reproduce this problem, though this problem also occurs for RKS computations).

Also note that currently, when installing numpy by conda (`defaults` channel), mkl=2023.1.0 will be also installed automatically, which is the problematic MKL version. To use mkl=2023.1.0 and reproduce problem, one may execuate bash command
```bash
conda create -n numpy-mkl23 python=3.10 numpy scipy h5py # mkl=2023.1.0 (for now this version is installed by default)
conda activate numpy-mkl23
export PYTHONPATH={/path/to/compiled/pyscf-2.4.0-or-master}:$PYTHONPATH
python problem.py
```

## Tracing the problem

Code where problem occurs (not meaning that the following code has bug there >.<):

https://github.com/pyscf/pyscf/blob/0bc590ffddec98eec5d2513a3d85274b7a791bc6/pyscf/lib/np_helper/npdot.c#L70-L89

where this code is called by `lib.dot`:

https://github.com/pyscf/pyscf/blob/0bc590ffddec98eec5d2513a3d85274b7a791bc6/pyscf/df/df_jk.py#L333-L335

For the computation which has problem, input matrix size is `(41, 9840, "T") * (9840, 41, "N") -> (41, 41)`, or `m = n = 41, k = 9840`. Values of input matrix `b` is all-zero, which has been verified (note that in `def _dgemm` in `numpy_helper.py` matrices `a, b` interchanged when actually calling C code `NPdgemm`).

Expected value of `cpriv` should be all-zero; however, actually `cpriv` is polluted after call of `dgemm_`. When printing information of `cpriv`, I found that this problem only occurs only on the 0-th thread (index of thread obtained by `int id_thread = omp_get_thread_num()`); value of `cpriv` in threads 1-7 is correct.

I also tried to zero-out `cpriv` before calling `dgemm_`; or defining zero-out `cpriv_buf` outside openmp parallel and then assign `double *cpriv = &cpriv_buf[(m*n+2) * id_thread]` inside openmp parallel. Neither approach gives correct result. 

I tried different BLAS (MKL 2023.2 of standalone oneAPI, compiled OpenBLAS 0.3.25, libblas and libopenblas-openmp shipped by RedHat or Ubuntu repositories) to compile PySCF and still find this problem exists. Since this problem is related to the BLAS linked with numpy (or shipped with python), not the BLAS that compiles PySCF, so I think `dgemm_` here refers to BLAS of numpy.

However, as noted above, this problem may not be reproduced when only a `lib.dot` called, with matrices of the same size as input. It's really tricky.

## Addition notes

This problem is so tricky, and bug of MKL is really rarely found, so I'm really not quite sure whether my analysis of this problem is correct.

Using `numpy.dot` instead of `lib.dot` will actually resolve this problem, for this specific case. But also note that I think that code in `lib.dot` and `NPdgemm` is correct; it's not a bug of PySCF.

I'm not very confident whether this problem could be reproduced by others. I only tried and verified on two different servers. And since I have not found decisive evidence that this problem is MKL 2023.1.0 bug, I'm not sure whether it is proper to report this problem to Intel, as well as by which means to report this problem to Intel.

Update: I tried to [post a question](https://community.intel.com/t5/Intel-oneAPI-Math-Kernel-Library/NumPy-with-MKL-2023-1-0-causes-PySCF-RI-K-integral-evaluated-by/m-p/1556176#M35647) in oneAPI MKL forum.

**Comment by ajz34** [comment]
I tried to [post a question](https://community.intel.com/t5/Intel-oneAPI-Math-Kernel-Library/NumPy-with-MKL-2023-1-0-causes-PySCF-RI-K-integral-evaluated-by/m-p/1556176#M35647) in oneAPI MKL forum.

**Comment by sunqm** [comment]
Is the complied pyscf extension linked to the libiomp or libgomp library?

**Comment by ajz34** [comment]
> Is the complied pyscf extension linked to the libiomp or libgomp library?

Seems that both libiomp and libgomp shows problem. Probably not related to OpenMP itself?

ldd information generated by `ldd pyscf/lib/libnp_helper.so`.

<details>
  <summary>Compiled by Intel oneAPI 2024.0 (CC=icx, CXX=icpx)</summary>

```
linux-vdso.so.1 =>  (0x00007ffe6979f000)
libmkl_intel_lp64.so.2 => /share/home/zyzhu/intel/oneapi/mkl/2024.0/lib/libmkl_intel_lp64.so.2 (0x00007f1d30dba000)
libmkl_intel_thread.so.2 => /share/home/zyzhu/intel/oneapi/mkl/2024.0/lib/libmkl_intel_thread.so.2 (0x00007f1d2e922000)
libmkl_core.so.2 => /share/home/zyzhu/intel/oneapi/mkl/2024.0/lib/libmkl_core.so.2 (0x00007f1d2a7f0000)
libiomp5.so => /share/home/zyzhu/intel/oneapi/compiler/2024.0/lib/libiomp5.so (0x00007f1d2a3a9000)
libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f1d2a18d000)
libimf.so => /share/home/zyzhu/intel/oneapi/compiler/2024.0/lib/libimf.so (0x00007f1d29d6f000)
libm.so.6 => /lib64/libm.so.6 (0x00007f1d29a6d000)
libdl.so.2 => /lib64/libdl.so.2 (0x00007f1d29869000)
libsvml.so => /share/home/zyzhu/intel/oneapi/compiler/2024.0/lib/libsvml.so (0x00007f1d28235000)
libirng.so => /share/home/zyzhu/intel/oneapi/compiler/2024.0/lib/libirng.so (0x00007f1d325d9000)
libintlc.so.5 => /share/home/zyzhu/intel/oneapi/compiler/2024.0/lib/libintlc.so.5 (0x00007f1d32577000)
libc.so.6 => /lib64/libc.so.6 (0x00007f1d27e67000)
librt.so.1 => /lib64/librt.so.1 (0x00007f1d27c5f000)
libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f1d27a49000)
/lib64/ld-linux-x86-64.so.2 (0x00007f1d324db000)
```

</details>

<details>
  <summary>Compiled by Intel gcc-12.2.0 with OpenBLAS (CC=gcc, CXX=g++)</summary>

```
linux-vdso.so.1 =>  (0x00007ffc350fd000)
libopenblas.so.0 => /share/home/zyzhu/Software/OpenBLAS-0.3.25/lib64/libopenblas.so.0 (0x00007fea055ae000)
libgomp.so.1 => /lib64/libgomp.so.1 (0x00007fea05388000)
libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fea0516c000)
libc.so.6 => /lib64/libc.so.6 (0x00007fea04d9e000)
libgfortran.so.5 => not found
libm.so.6 => /lib64/libm.so.6 (0x00007fea04a9c000)
libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fea04886000)
libquadmath.so.0 => /lib64/libquadmath.so.0 (0x00007fea0464a000)
/lib64/ld-linux-x86-64.so.2 (0x00007fea06a96000)
```

</details>

<details>
  <summary>interlude when compiled by icx</summary>

Following error message is unrelated to this topic.

https://github.com/pyscf/pyscf/blob/0bc590ffddec98eec5d2513a3d85274b7a791bc6/pyscf/lib/mcscf/fci_contract.c#L607C16-L607C21

fci_contract.c:601:16: error: call to undeclared function 'ffsll'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]

</details>

**Comment by ajz34** [comment]
# Docker workflow to reproduce this issue

Also refers to file [docker_issue_2004.zip](https://github.com/pyscf/pyscf/files/13758028/docker_issue_2004.zip) for build instructions of docker.

## Use image that is already on hub.docker.com

Before entering docker interactive shell,

```bash
docker pull ajz34/issue2004
docker run -it --cpus=8 ajz34/issue2004:latest
```

After entering docker interactive shell,

```bash
conda activate numpy-mkl23
python problem.py
```

Correct value of output matrix is all zero. On Intel CPUs with more than 8 physical cores, the result could be random non-zero values.

Please make sure that CPU is Intel, and CPU has more than 8 physical cores, to reproduce this issue.

## Build this docker image

To build this docker image,

```bash
docker build -t issue2004 .
```

This involves pull of docker and github, so make sure network is available.

Also make sure `problem.py` file is included along `Dockerfile`.

**Comment by sunqm** [comment]
It is caused by MKL threads. You can set the environment MKL_NUM_THREADS=1 or use the threadpoolctl library to manage MKL threads at runtime.

In the conda release of the mkl-numpy library, a custom module `mkl` is imported by numpy. This `mkl` module pollutes the symbols of BLAS functions.  You can find in the `__init__.py` of the `mkl` module that MKL library is load with `sys.setdlopenflags(self.saved_rtld | ctypes.RTLD_GLOBAL)`. This setting makes the BLAS functions from MKL be the global symbols. As a result, the dgemm_ in pyscf extensions is actually bound to MKL library, although you see libopenblas in ldd command.

PySCF extensions are designed to work with single-thread BLAS only. The MKL you used is controlled by the runtime driver libmkl_rt.so, which calls multithreading dgemm_ functions by default. You may refer to MKL manual how to avoid the mkl rt driver when compiling pyscf extensions.

**Comment by ajz34** [comment]
This discussion is really helpful😄 Though I'll argue with some points.

------

> It is caused by MKL threads.

It is correct. This problem is very likely due to MKL 2023.1's bug of dynamically changing the number of OpenMP threads. And pointing out the way how `mkl` module works (introduced by `mkl-service`) is helpful.

I guess that MKL should be designed to be able to run `dgemm_` single-threaded (or properly assign the number of threads) within `omp parallel`, when [`mkl_get_dynamic`](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-1/mkl-get-dynamic.html) gives True value.

This could tip down by printing [`mkl_get_max_threads()`](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-1/mkl-get-max-threads.html) inside OMP parallel region in [npdot.c#L70-L89](https://github.com/pyscf/pyscf/blob/0bc590ffddec98eec5d2513a3d85274b7a791bc6/pyscf/lib/np_helper/npdot.c#L70-L89). The correct behavior is that when given 8 threads available to executable, every OMP parallel region uses only one thread; while in this erroneous case, 0-th parallel region consumes 8 threads, and 1st-7th parallel region consumes 1 thread. The 0-th parallel region is racing with other parallel regions.

This erroneous behavior would also occur when basis set is set to cc-pVDZ, or auxiliary basis set set to cc-pVDZ-jkfit, though sometimes this kind of parallel race may still give correct values by chance. However, in the case of aug-cc-pVDZ with ETB auxiliary basis, wrong results could be reproduced.

------

Some argues follows.

> You can set the environment MKL_NUM_THREADS=1 or use the threadpoolctl library to manage MKL threads at runtime.

Setting `MKL_NUM_THREADS=1` will also disable multi-threading of numpy functions, if numpy is linked with MKL. So disabling this environment variable could incur great efficiency loss on multi-threading task, if numpy functions are also computational bottlenecks.

> which calls multithreading dgemm_ functions by default

For MKL, as stated before, though MKL has bug here, I think that by design, MKL will properly assign number of threads for `dgemm_` in OMP parallel regions. So without very delicated fine tuning of OMP parallel controls, `dgemm_` should be called in single-threaded fashion for MKL.

For OpenBLAS, this library does not dynamically assign number of threads inside OMP parallel regions. However, if compiled with OpenMP (cmake `-DUSE_OPENMP=1`), it will use only one thread if `dgemm_` is called inside OMP parallel regions. Also see [common_thread.h L141-L154](https://github.com/OpenMathLib/OpenBLAS/blob/develop/common_thread.h#L141-L154) of OpenBLAS. But also note that OpenBLAS does not utilize OpenMP with default compile flags.

**Comment by ajz34** [comment]
Since MKL 2024 is available, and currently MKL 2023.2 shipped by channel `conda-forge` seems not showing erroneous results in this specific case, so perhaps this will not be an issue as months passed. But as some complaints, I've already encountered this issue, mostly to radicals that are hard to obtain stable wavefunctions with large diffuse basis sets, for months and wasted days analyzing data with wrong values. So I'm curious on why this happens. And though MKL is really fast on Intel's own CPUs, I will try to be more cautious when using MKL with numpy.

I believe this is surely a bug of MKL 2023.1. Still, I haven't worked out a minimal case that does not involve PySCF, that also shows MKL threading control is flawed. I will update [post in MKL forum](https://community.intel.com/t5/Intel-oneAPI-Math-Kernel-Library/NumPy-with-MKL-2023-1-0-causes-PySCF-RI-K-integral-evaluated-by/m-p/1556176) if this problem could be reproduced with less code efforts. I may work on this issue in spare time.
